Add event tracing and ETDumps to executor_runner

benkli01 opened the pull request:
- Enabled via EXECUTORCH_ENABLE_EVENT_TRACER
- Add flag 'etdump_path' to specify the file path for the ETDump file
- Add flag 'num_executions' for number of iterations to run
- Create and pass event tracer 'ETDumpGen'
- Save ETDump to disk
- Update docs to reflect the changes

benkli01 committed:
Add event tracing and ETDumps to executor_runner

- Enabled via EXECUTORCH_ENABLE_EVENT_TRACER
- Add flag 'etdump_path' to specify the file path for the ETDump file
- Add flag 'num_executions' for number of iterations to run
- Create and pass event tracer 'ETDumpGen'
- Save ETDump to disk
- Update docs to reflect the changes

Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>
Change-Id: I876d5138455d1b04fba9af4016d8341e8866f9c0

benkli01 commented:
@pytorchbot label 'partner:arm'

benkli01 commented:
@pytorchbot label 'partner: arm'

benkli01 commented:
The failing test looks unrelated to my changes.

digantdesai commented on docs/source/native-delegates-executorch-xnnpack-delegate.md:
```diff
@@ -74,7 +74,7 @@ Since weight packing creates an extra copy of the weights inside XNNPACK, We fre
 
 #### **Profiling**
-We have enabled basic profiling for XNNPACK delegate that can be enabled with the following compiler flag `-DENABLE_XNNPACK_PROFILING`. With ExecuTorch's SDK integration, you can also now use the SDK tools to profile the model. You can follow the steps in [Using the ExecuTorch SDK to Profile a Model](./tutorials/sdk-integration-tutorial) on how to profile ExecuTorch models and use SDK's Inspector ...
+We have enabled basic profiling for XNNPACK delegate that can be enabled with the following compiler flag `-DEXECUTORCH_ENABLE_EVENT_TRACER`. With ExecuTorch's SDK integration, you can also now use the SDK tools to profile the model. You can follow the steps in [Using the ExecuTorch SDK to Profile a Model](./tutorials/sdk-integration-tutorial) on how to profile ExecuTorch models and use SDK's Ins ...
```
`ENABLE_XNNPACK_PROFILING` is still available for more detailed profiling like operator timing. We should reflect that.

benkli01 replied:
Good point. I updated the docs. 👍

digantdesai commented on docs/source/tutorial-xnnpack-delegate-lowering.md:
```diff
@@ -171,3 +171,6 @@ Now you should be able to find the executable built at `./cmake-out/backends/xnn
 You can build the XNNPACK backend [CMake target](https://github.com/pytorch/executorch/blob/main/backends/xnnpack/CMakeLists.txt#L83), and link it with your application binary such as an Android or iOS application. For more information on this you may take a look at this [resource](demo-apps-android.md) next.
+
+## Profiling
+To enable profiling in the `xnn_executor_runner` pass the flags `-DEXECUTORCH_ENABLE_EVENT_TRACER=ON` and `-DEXECUTORCH_BUILD_SDK=ON` to the build command. This will enable ETDump generation when running the inference and enables command line flags for profiling (see `xnn_executor_runner --help` for details).
```
same here

digantdesai commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -113,8 +114,13 @@ if(NOT CMAKE_TOOLCHAIN_FILE MATCHES ".*(iOS|ios\.toolchain)\.cmake$")
   add_executable(xnn_executor_runner ${_xnn_executor_runner__srcs})
+
+  if(EXECUTORCH_ENABLE_EVENT_TRACER)
+    list(APPEND xnn_executor_runner_libs etdump)
```
I guess this never worked for OSS. Thanks. We should also consider adding `ENABLE_XNNPACK_PROFILING` CMake flag here.

benkli01 replied:
The flag `ENABLE_XNNPACK_PROFILING` is not used by the executor_runner to enable etdumps, so I thought it should be fine to just use the `EXECUTORCH_ENABLE_EVENT_TRACER`.

digantdesai replied:
Yeah not an issue with your PR.

digantdesai commented on examples/portable/executor_runner/executor_runner.cpp:
```diff
@@ -29,13 +30,23 @@
+    etdump_path,
+    "model.etdump",
+    "If ETDump generation is enabled an ETDump will be written out to this path.");
+DEFINE_uint32(num_executions, 10, "Number of times to run the model.");
```
Do you want to make this available more generally outside EVENT_TRACER with default value = 1?

digantdesai reviewed (approved):
Looks good to me. Left some comments. I assume you tried running this with both portable and XNNPACK, with and without this flag and it works OK?

Do you want to add a test with this flag enabled just so we can catch any build issues in this part of the logic?

benkli01 committed:
Minor changes to event tracing in executor_runner

- Make flag 'num_executions' available in the executor_runner
  irrespective of the event tracing
- Update docs to explain usage of 'ENABLE_XNNPACK_PROFILING' for
  additional profiling info

Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>
Change-Id: I35abbd2d913880cb129bddb80514992f4dd84004

benkli01 commented:
Thanks for the review @digantdesai ! I pushed a patch with some changes.

> I assume you tried running this with both portable and XNNPACK, with and without this flag and it works OK?

Yes, I didn't see any problems.

> Do you want to add a test with this flag enabled just so we can catch any build issues in this part of the logic?

Sure. Where would you add this test? I haven't found an obvious place.

GregoryComer commented:
Hey @benkli01, I ended up needing to back out this change due to it breaking the test-llama-runner-qnn-linux job (see https://github.com/pytorch/executorch/actions/runs/10398365231/job/28795542762). I apologize for the inconvenience, it didn't get caught on the PR. I'm happy to help re-land this, as this is a change that we want to keep. Thanks.

digantdesai commented:
cc @cccclai

benkli01 commented:
Re-opened with https://github.com/pytorch/executorch/pull/5027 as this change was reverted.
